src/OpalCompiler-Core/OCBytecodeToASTCache.class.st

```diff
@@ -0,0 +1,95 @@
+"
+I cache a mapping between possible bytecode offsets and the AST nodes they correspond to for a given compiled method.
+
+Instanciate me using my class method generateForCompiledMethod: and give me as parameter a compiled method.
+
+Use me through the node access API method nodeForPC: and give me a program counter as parameter.
+
+I store: 
+- firstBcOffset: The first bytecode pc. If you try to access a pc below that first pc, I return the method node.
+- lastBcOffset: The last bytecode pc. If you try to access a pc after this last pc, I return the node associated with the last pc.
+- bcToASTMap: A map associating each possible pc between firstBcOffset and lastBcOffset and the corresponding ast node.
+- the methode node.
+"
+Class {
+	#name : #OCBytecodeToASTCache,
+	#superclass : #Object,
+	#instVars : [
+		'firstBcOffset',
+		'lastBcOffset',
+		'bcToASTMap',
+		'methodNode'
+	],
+	#category : #'OpalCompiler-Core-Bytecode'
+}
+
+{ #category : #initialization }
+OCBytecodeToASTCache class >> generateForMethodNode: aNode [
+	^self new generateForMethodNode: aNode
+]
+
+{ #category : #accessing }
+OCBytecodeToASTCache >> bcToASTMap [
+	^ bcToASTMap
+]
+
+{ #category : #private }
+OCBytecodeToASTCache >> fillMissingBCOffsetsWithLastBCOffsetNodes [
+	"It happens that different bytecode offsets map to the same AST node.
+	These cases are detected when there is no node mapped between, for example, bcOffset 46 and bcOffset 50.
+	In that case, we take every possible bytecode index between 46 and 50 (i.e., 47, 48, 49),
+	and we map them to the same node as the last mapped bytecode offset, here 46."
+
+	| sortedBCOffsets |
+	sortedBCOffsets := bcToASTMap keys asSortedCollection.
+	1 to: sortedBCOffsets size - 1 do: [ :index | 
+		| bcAtIndex bcAtNextIndex |
+		bcAtIndex := sortedBCOffsets at: index.
+		bcAtNextIndex := sortedBCOffsets at: index + 1.
+		bcAtIndex < bcAtNextIndex ifTrue: [ 
+			bcAtIndex to: bcAtNextIndex - 1 do: [ :i | 
+			bcToASTMap at: i put: (bcToASTMap at: bcAtIndex) ] ] ]
+]
+
+{ #category : #accessing }
+OCBytecodeToASTCache >> firstBcOffset [
+	^ firstBcOffset
+]
+
+{ #category : #initialization }
+OCBytecodeToASTCache >> generateForMethodNode: aNode [
+	| methodIR currentBcOffset |
+	methodNode := aNode.
+	methodIR := methodNode ir.
+	bcToASTMap := Dictionary new.
+	firstBcOffset := methodNode compiledMethod initialPC.
+	
+	"if there is a primitive, we need to record the offset"
+	(methodIR irPrimitive num ~= 0) ifTrue: [ bcToASTMap at: firstBcOffset put: methodNode].
```

Comment:
This case with primitives is not covered by a test. 
Could you add one?

Could you give more details about that case? I would like to document it in the compiler booklet.

Reply:
There is #testPrimitiveMethodSourceNodeAtInitialPC which failed without this.

So the thing is that primitives are modelled with a bytecode, but there irPrimitive is not part of the other IR instructions, but instead the IRMethod has an ivar that references is.

This is something we could actually change.

Reply:
(but not as part of this PR)